// tiktok-live-connector/dist/lib/platforms/douyinConfig.js
"use strict";

module.exports = {
  NAME: 'douyin',
  TIKTOK_URL_WEB: 'https://live.douyin.com/',
  TIKTOK_URL_LIVE_PATH: '%uniqueId%',
  TIKTOK_URL_WEBCAST: 'https://live.douyin.com/webcast/',
  TIKTOK_HTTP_ORIGIN: 'https://live.douyin.com',
  DEFAULT_CLIENT_PARAMS: {
    'X-Bogus': 'DFSzswVuIBUANjcPSlMPr5YklT6b',
    aid: '6383',
    cursor: '',
    device_platform: 'web',
    did_rule: '3',
    identity: 'audience',
    language: 'en-US',
    last_rtt: '295',
    live_id: '1',
    resp_content_type: 'protobuf',
    version_code: '9999'
  },
  DEFAULT_REQUEST_HEADERS: {
    Connection: 'keep-alive',
    'Cache-Control': 'max-age=0',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36',
    Accept: 'text/html,application/json,application/protobuf',
    Referer: 'https://live.douyin.com/',
    Origin: 'https://live.douyin.com',
    'Accept-Language': 'en-US,en;q=0.9',
    'Accept-Encoding': 'gzip, deflate'
  }
};
